qutip/core/data/block_operations.pyx

```diff
@@ -0,0 +1,423 @@
+        while idx < num_ops:
+            if block_rows[idx] != row:
+                break
+            idx += 1
+        # now the operators in the current row have ids (prev_idx, ..., idx-1)
+
+        for op_row in range(block_heights[row]):
+            for i in range(prev_idx, idx):
+                op = <CSR>blocks[i]
+                if csr.nnz(op) == 0:
+                    # empty CSR matrices have uninitialized row_index entries.
+                    # it's unclear whether users should ever see such matrixes
+                    # but we support them here anyway.
+                    continue
+
+                column = block_cols[i]
+                op_row_start = op.row_index[op_row]
+                op_row_end = op.row_index[op_row + 1]
+                op_row_len = op_row_end - op_row_start
+                for counter in range(op_row_len):
+                    out.col_index[end + counter] = (
+                        op.col_index[op_row_start + counter] +
+                        col_pos[column]
+                    )
+                    out.data[end + counter] =\
+                        op.data[op_row_start + counter]
+                end += op_row_len
+            out.row_index[row_pos[row] + op_row + 1] = end
+
+    return out
+
+
+cpdef Dense slice_dense(Dense data,
+                        base.idxint row_start, base.idxint row_stop,
+                        base.idxint col_start, base.idxint col_stop):
+    if (
+        row_start < 0 or col_start < 0
+        or row_stop > data.shape[0] or col_stop > data.shape[1]
+        or row_start >= row_stop or col_start >= col_stop
+    ):
+        raise IndexError("Slice indices are out of bounds.")
+    cdef cnp.ndarray array = data.as_ndarray()
+    return Dense(array[row_start:row_stop, col_start:col_stop], copy=True)
+
+
+cpdef CSR slice_csr(CSR data,
+                    base.idxint row_start, base.idxint row_stop,
+                    base.idxint col_start, base.idxint col_stop):
+    if (
+        row_start < 0 or col_start < 0
+        or row_stop > data.shape[0] or col_stop > data.shape[1]
+        or row_start >= row_stop or col_start >= col_stop
+    ):
+        raise IndexError("Slice indices are out of bounds.")
+    scipy = data.as_scipy()
+    return CSR(scipy[row_start:row_stop, col_start:col_stop], copy=True)
+
+
+cpdef Dense insert_block_dense(Data data, Dense block,
+                               base.idxint above, base.idxint before):
+    cdef base.idxint data_height, data_width, block_height, block_width
+
+    data_height, data_width = data.shape
+    block_height, block_width = block.shape
+    if (
+        above < 0 or before < 0
+        or above + block_height > data_height
+        or before + block_width > data_width
+    ):
+        raise IndexError("Cannot insert block into data: doesn't fit.")
+
+    cdef cnp.ndarray data_array = data.to_array()  # copies
+    cdef cnp.ndarray block_array = block.as_ndarray()  # doesn't copy
```

Comment:
We could easily also support `block.to_array()` here (e.g. via checking the block type) and then `insert_block_dense` could support `Data, Data, Dense`.